File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/schedule/ScheduleRebootViewModel.kt
package com.openu60.feature.router.schedule

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.ScheduleRebootConfig
import com.openu60.core.model.ScheduleRebootParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ScheduleRebootState(
    val config: ScheduleRebootConfig = ScheduleRebootConfig.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class ScheduleRebootViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(ScheduleRebootState())
    val state: StateFlow<ScheduleRebootState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSON("/api/device/schedule-reboot")
                val config = ScheduleRebootParser.parse(data)
                _state.value = _state.value.copy(config = config, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun updateConfig(config: ScheduleRebootConfig) {
        _state.value = _state.value.copy(config = config)
    }

    fun save() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val config = _state.value.config
                agentClient.putJSON("/api/device/schedule-reboot", mapOf(
                    "auto_reboot_enable" to if (config.enabled) "1" else "0",
                    "auto_reboot_time" to config.time,
                    "auto_reboot_days" to config.days,
                ))
                _state.value = _state.value.copy(isLoading = false, message = "Saved", messageIsError = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) save() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}

File: mobile/ios/OpenU60/Features/RouterSettings/Schedule/ScheduleRebootViewModel.swift
import SwiftUI

@Observable
@MainActor
final class ScheduleRebootViewModel {
    var config: ScheduleRebootConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    var editEnabled: Bool = false
    var editTime: Date = Calendar.current.date(from: DateComponents(hour: 3, minute: 0)) ?? Date()
    var editDays: Set<String> = []

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let data = try await client.getJSON("/api/modem/schedule-reboot")
            config = ScheduleRebootParser.parse(data)
            editEnabled = config.enabled
            editTime = parseTime(config.time)
            editDays = Set(config.days.split(separator: ",").map(String.init))
        } catch {
            showMessage("Failed to load schedule: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func apply() async {
        isLoading = true

        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"
        let timeStr = formatter.string(from: editTime)
        let daysStr = editDays.sorted().joined(separator: ",")

        do {
            let _ = try await client.putJSON("/api/modem/schedule-reboot", body: [
                "auto_reboot_enable": editEnabled ? "1" : "0",
                "auto_reboot_time": timeStr,
                "auto_reboot_days": daysStr
            ])
            showMessage("Schedule updated", isError: false)
            config = ScheduleRebootConfig(enabled: editEnabled, time: timeStr, days: daysStr)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func parseTime(_ timeStr: String) -> Date {
        let parts = timeStr.split(separator: ":")
        guard parts.count == 2,
              let hour = Int(parts[0]),
              let minute = Int(parts[1]) else {
            return Calendar.current.date(from: DateComponents(hour: 3, minute: 0)) ?? Date()
        }
        return Calendar.current.date(from: DateComponents(hour: hour, minute: minute)) ?? Date()
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}
